api/src/main/java/org/eclipse/microprofile/metrics/MetadataBuilder.java

```diff
@@ -0,0 +1,203 @@
+     */
+    public MetadataBuilder withDescription(String description) {
+        this.description = Objects.requireNonNull(description, "description is required");
+        return this;
+    }
+
+    /**
+     * Sets the type
+     *
+     * @param type the name
+     * @return the builder instance
+     * @throws NullPointerException when type is null
+     */
+    public MetadataBuilder withType(MetricType type) {
+        this.type = Objects.requireNonNull(type, "type is required");
+        return this;
+    }
+
+    /**
+     * Sets the unit
+     *
+     * @param unit the name
+     * @return the builder instance
+     * @throws NullPointerException when unit is null
+     */
+    public MetadataBuilder withUnit(String unit) {
+        this.unit = Objects.requireNonNull(unit, "unit is required");
+        return this;
+    }
+
+    /**
+     * Sets the reusable to {@link Boolean#TRUE}
+     *
+     * @return the builder instance
+     */
+    public MetadataBuilder reusable() {
+        this.reusable = true;
+        return this;
+    }
+
+    /**
+     * Sets the reusable to {@link Boolean#FALSE}
+     *
+     * @return the builder instance
+     */
+    public MetadataBuilder notReusable() {
+        this.reusable = false;
+        return this;
+    }
+
+    /**
+     * Add multiple tags delimited by commas.
+     * The format must be in the form 'key1=value1, key2=value2'.
+     * This method will call {@link #addTag(String)} on each tag.
+     *
+     * @param tagsString a string containing multiple tags
+     * @return this instance with new tags
+     */
+    public MetadataBuilder addTags(String tagsString) {
+        if (tagsString == null || tagsString.isEmpty()) {
+            return this;
+        }
+        String[] singleTags = tagsString.split(",");
+        Stream.of(singleTags).map(String::trim).forEach(this::addTag);
+        return this;
+    }
+
+    /**
+     * Add one single tag with the format: 'key=value'. If the input is empty or does
+     * not contain a '=' sign, the entry is ignored.
+     *
+     * @param kvString Input string
+     * @return this instance with new tag
+     */
+    public MetadataBuilder addTag(String kvString) {
+        if (kvString == null || kvString.isEmpty() || !kvString.contains("=")) {
+            return this;
+        }
+        tags.put(kvString.substring(0, kvString.indexOf("=")), kvString.substring(kvString.indexOf("=") + 1));
+        return this;
+    }
+
+    /**
+     * @return
+     * @throws IllegalStateException when either name is null
+     */
+    public Metadata build() {
+        if (Objects.isNull(name)) {
+            throw new IllegalStateException("Name is required");
+        }
+        switch (type) {
+            case HISTOGRAM:
+            case TIMER:
+            case METERED:
+            if(Objects.isNull(unit) || MetricUnits.NONE.equals(unit)) {
+                throw new IllegalStateException("To this type of Metadata the unit is required");
```

Comment:
In the spec...
> Counter metrics are considered dimensionless. The implementation must not append the unit name to the metric name and must not scale the value.

> Meter and Timer have fixed units as described below regardless of the unit value in the Metadata.

So the logic should be:
Gauges, Histograms - units required
Counters, Timers, Meters - no units allowed